### ISSUE TITLE
isAuthenticated and express middleware issue

### ISSUE THREAD

**Issue by Alejogb1** [Alejogb1/NodeApp#4]
I'm  trying to create  a middleware that redirects me to _dashboard_ whenever the user is logged in. I use the function "_isAuthenticated_" to check whether the user is authenticated or not. The problem is that whenever the middleware tries to check this last function, it gives me false and redirects me to "/#" when it should be "/dashboard"

**Comment by Nommer101** [comment]
https://github.com/Alejogb1/NodeApp/blob/55534b0991ef50df71d49431af79717b818c0021/routes/auth.js#L13

Isn't that line your issue on failure redirect to / which is your login page?

**Comment by Nommer101** [comment]
https://github.com/Alejogb1/NodeApp/blob/d46a8e19b638aafee0d0ae9ad09ad6084d42dff3/config/passport.js#L9
Callback link needs to be absolute address to your site. You are telling google to goto google after it authenticated.

**Comment by Nommer101** [comment]
https://github.com/Alejogb1/NodeApp/blob/d46a8e19b638aafee0d0ae9ad09ad6084d42dff3/routes/auth.js#L15

You need to create an express session here and set it in cookies or however you want to handle sessions.
Your client needs to know how they will be identified as logged in by the server.

See https://github.com/Alejogb1/NodeApp/pull/5

**Comment by Alejogb1** [comment]
> https://github.com/Alejogb1/NodeApp/blob/d46a8e19b638aafee0d0ae9ad09ad6084d42dff3/config/passport.js#L9
> 
> 
> Callback link needs to be absolute address to your site. You are telling google to goto google after it authenticated.

I've changed as you've said and I'm still having the same issue. Thank you very much, tho.

**Comment by Nommer101** [comment]
Your issue is your domain needs to be approved allowed on google dev console. If you do not get the google error on auth link then please let me know where the error is.

**Comment by Alejogb1** [comment]
![Capturaas](https://user-images.githubusercontent.com/52103963/91449946-1aa1e000-e852-11ea-9115-534ff79d2907.PNG)
No, I've had that issue before and then I solved it by adding on the Url's allowed, the "http://localhost:3000/dashboard" domain. If I wouldn't have been allowed I would receive a message when trying to authenticate on the Google login window but that's not the case.

**Comment by Alejogb1** [comment]
I can't see which is the error. In my opinion, whenever I login, **the isAuthenticated() gives me false when it should be true**, so it doesn't redirects me to _/dashboard_ (I will leave some images from a debugging session I did). 
      On the first image, I don't get where I should see if the **isAuthenticated() function is returning me a false value** (I don't know where to look at). 
      And on the second photo,  I selected with red ink that as the isAuthenticated doesn't return me a true value, so it could to next(). Aversely, It **goes the else section and redirects me to "/".** 
![image (1)](https://user-images.githubusercontent.com/52103963/91451516-e0d1d900-e853-11ea-97e2-72816849986e.png)
![image (2)](https://user-images.githubusercontent.com/52103963/91451521-e16a6f80-e853-11ea-9bba-3d47e014001d.png)

**Comment by Nommer101** [comment]
You did not pull the code that was merged.
https://github.com/Alejogb1/NodeApp/blob/e2d85a740f5f48192771bb8525b91422aa4f0b70/middleware/auth.js#L3

**Comment by Alejogb1** [comment]
This is the code I have in my local file, and it is the same from the repo, so I don't have to pull nothing. I appreciate your help.
`module.exports = {
    ensureAuth: function (req, res, next) {
        if (req.session.isAuthenticated) {
            return next();
        } else {
            res.redirect("/");
        }
    },
    ensureGuest: function (req, res, next) {
        if (!req.session.isAuthenticated) {
            return next();
        } else {
            res.redirect("/dashboard");
        }
    },
}`

**Comment by Nommer101** [comment]
Hey, I tested it and it works.

Your screenshot shows old code thou. That is why I said you need to pull. It should work unless its failing with database insert.

I just killed the database entry and returned the profile instead.

```
module.exports = function (passport) {
  passport.use(new GoogleStrategy({
    clientID: process.env.GOOGLE_CLIENT_ID,
    clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    callbackURL: "http://localhost:3000/auth/google/callback"
  },
    async (accessToken, refreshToken, profile, done) => {
      console.log(profile);
      done(null, profile);
    }
  ))
  passport.serializeUser((user, done) => { // Grab User Infor and "done"
    done(null, user.id); // user.id is MD ID, NOT google's id
  });

  passport.deserializeUser((id, done) => {
    //User.findById(id, (err, user) => done(err, user));
  });
}
```

**Comment by Alejogb1** [comment]
Hey, when you say it worked, did you get redirected to the /dashboard when logging in? Also, in the code you've sent shouldn't be there the dashboard callback URL instead of the callback one?

**Comment by Alejogb1** [comment]
I receive this message when entering to localhost: 
```
Session {
  cookie: { path: '/', _expires: null, originalMaxAge: null, httpOnly: true }
}
```

**Comment by Nommer101** [comment]
https://github.com/Nommer101/NodeApp
Pull that and see if it works for you. It should

**Comment by Nommer101** [comment]
Still having issues with this?

**Comment by Alejogb1** [comment]
Yes

**Comment by Alejogb1** [comment]
> https://github.com/Nommer101/NodeApp
> Pull that and see if it works for you. It should

When you run the code in your local repo, does login works fine and redirects you to the "/dashobard" page?

**Comment by Nommer101** [comment]
> When you run the code in your local repo, does login works fine and redirects you to the "/dashobard" page?
Yes, I used my branch but killed the database and it worked fine. It was using your api key for the google authentication. I removed your app from my account afterwards.
